segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -0,0 +1,510 @@
+    abstract protected TransactionData read(String key) throws StorageMetadataException;
+
+    /**
+     * Writes transaction data from a given list to the metadata store.
+     * @param dataList List of transaction data to write.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @VisibleForTesting
+    abstract protected void writeAll(Collection<TransactionData> dataList) throws StorageMetadataException;
+
+    /**
+     * Updates existing metadata.
+     *
+     * @param txn Transaction.
+     * @param metadata metadata record.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    public void update(MetadataTransaction txn, StorageMetadata metadata) throws StorageMetadataException  {
+        Preconditions.checkArgument(null != txn);
+        Preconditions.checkArgument(null != metadata);
+        Preconditions.checkArgument(null != metadata.getKey());
+        Map<String, TransactionData> txnData = txn.getData();
+
+        String key = metadata.getKey();
+
+        TransactionData data = TransactionData.builder().key(key).build();
+        TransactionData oldData = txnData.putIfAbsent(key, data);
+        if (null != oldData) {
+            data = oldData;
+        }
+        data.setValue(metadata);
+        data.setPersisted(false);
+        Preconditions.checkState(txn.getVersion() >= data.getVersion());
+        data.setVersion(txn.getVersion());
+    }
+
+    /**
+     * Marks given record as pinned.
+     *
+     * @param txn Transaction.
+     * @param metadata metadata record.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    public void markPinned(MetadataTransaction txn, StorageMetadata metadata) throws StorageMetadataException  {
+        Preconditions.checkArgument(null != txn);
+        Preconditions.checkArgument(null != metadata);
+        Map<String, TransactionData> txnData = txn.getData();
+        String key = metadata.getKey();
+
+        TransactionData data = TransactionData.builder().key(key).build();
+        TransactionData oldData = txnData.putIfAbsent(key, data);
+        if (null != oldData) {
+            data = oldData;
+        }
+
+        data.setValue(metadata);
+        data.setPinned(true);
+        data.setVersion(txn.getVersion());
+
+        pinnedKeys.add(metadata.getKey());
+    }
+
+    /**
+     * Creates a new metadata record.
+     *
+     * @param txn Transaction.
+     * @param metadata  metadata record.
+     * @throws StorageMetadataException Exception related to storage metadata operations.
+     */
+    @Override
+    public void create(MetadataTransaction txn, StorageMetadata metadata) throws StorageMetadataException  {
+        Preconditions.checkArgument(null != txn);
+        Preconditions.checkArgument(null != metadata);
+        Preconditions.checkArgument(null != metadata.getKey());
```

Comment:
Didn't you validate this inside the metadata object?

Reply:
no that was in other places.